lib/needlessDisables.js

```diff
@@ -1,95 +1,92 @@
-		const unused = { source: result.source || '', ranges: [] };
-
-		/** @type {{[ruleName: string]: Array<RangeType>}} */
-		const rangeData = _.cloneDeep(result._postcssResult.stylelint.disabledRanges);
+		/** @type {{[ruleName: string]: Array<DisabledRange>}} */
+		const rangeData = result._postcssResult.stylelint.disabledRanges;
 
 		if (!rangeData) {
 			return;
 		}
 
 		const disabledWarnings = result._postcssResult.stylelint.disabledWarnings || [];
 
+		// A map from `stylelint-disable` comments to the set of rules that
+		// are usefully disabled by each comment. We track this
+		// comment-by-comment rather than range-by-range because ranges that
+		// disable *all* rules are duplicated for each rule they apply to in
+		// practice.
+		/** @type {Map<PostcssComment, Set<string>>}} */
+		const usefulDisables = new Map();
+
 		disabledWarnings.forEach((warning) => {
 			const rule = warning.rule;
-
 			const ruleRanges = rangeData[rule];
 
 			if (ruleRanges) {
-				// Back to front so we get the *last* range that applies to the warning
-				for (const range of ruleRanges.reverse()) {
+				for (const range of ruleRanges) {
 					if (isWarningInRange(warning, range)) {
-						range.used = true;
-
-						return;
+						putIfAbsent(usefulDisables, range.comment, () => new Set()).add(rule);
 					}
 				}
 			}
 
-			for (const range of rangeData.all.reverse()) {
+			for (const range of rangeData.all) {
 				if (isWarningInRange(warning, range)) {
-					range.used = true;
-
-					return;
+					putIfAbsent(usefulDisables, range.comment, () => new Set()).add(rule);
 				}
 			}
 		});
 
-		Object.keys(rangeData).forEach((rule) => {
-			rangeData[rule].forEach((range) => {
-				// Is an equivalent range already marked as unused?
-				const alreadyMarkedUnused = unused.ranges.find((unusedRange) => {
-					return unusedRange.start === range.start && unusedRange.end === range.end;
-				});
+		// Get rid of the duplicated ranges for each `all` rule. We only care
+		// if the entire `all` rule is useful as a whole or not.
+		for (const range of rangeData.all) {
+			for (const [rule, ranges] of Object.entries(rangeData)) {
+				if (rule === 'all') continue;
 
-				// If this range is unused and no equivalent is marked,
-				// mark this range as unused
-				if (!range.used && !alreadyMarkedUnused) {
-					unused.ranges.push({
-						rule,
-						start: range.start,
-						end: range.end,
-						unusedRule: rule,
-					});
-				}
+				_.remove(ranges, (otherRange) => range.comment === otherRange.comment);
+			}
+		}
 
-				// If this range is used but an equivalent has been marked as unused,
-				// remove that equivalent. This can happen because of the duplication
-				// of ranges in rule-specific range sets and the "all" range set
-				if (range.used && alreadyMarkedUnused) {
-					_.remove(unused.ranges, alreadyMarkedUnused);
-				}
+		Object.entries(rangeData).forEach(([rule, ranges]) => {
```

Comment:
let's not mix `forEach` with `for..of`. btw you can optimize `const entries = Object.entries(rangeData)` and use `entries` identifier both on line 67 and line 60.

Reply:
Done.